lib/screens/screens_by_alert/gen_server.ex

```diff
@@ -39,23 +49,133 @@ defmodule Screens.ScreensByAlert.GenServer do
     GenServer.call(pid, {:get_screens_last_updated, screen_id})
   end
 
+  ### Server
+
+  @impl GenServer
+  def init(opts) do
+    {:ok,
+     %__MODULE__{
+       screens_by_alert_ttl_seconds: opts[:screens_by_alert_ttl_seconds],
+       screens_last_updated_ttl_seconds: opts[:screens_last_updated_ttl_seconds]
+     }}
+  end
+
+  @impl GenServer
+  def handle_cast({:put_data, screen_id, alert_ids}, %__MODULE__{} = state) do
+    updated_screens_by_alert =
+      alert_ids
+      |> Enum.map(fn alert_id ->
+        # Create a new timer to expire object in #{screens_by_alert_ttl_seconds} seconds
+        reference =
+          Process.send_after(
+            self(),
+            {:expire_alert, alert_id},
+            state.screens_by_alert_ttl_seconds * 1000
+          )
+
+        {alert_id, %{screen_ids: MapSet.new([screen_id]), timer_reference: reference}}
+      end)
+      |> Map.new()
+      # Combine list of screen_ids if alert already exists.
+      # Otherwise, create a new alert.
+      |> Map.merge(state.screens_by_alert, fn
+        _,
+        %{screen_ids: screens1, timer_reference: new_timer},
+        %{screen_ids: screens2, timer_reference: old_timer} ->
+          _ = Process.cancel_timer(old_timer, async: true, info: false)
+          %{screen_ids: MapSet.union(screens1, screens2), timer_reference: new_timer}
```

Comment:
Wait—I might have missed something here. As long as we keep getting updates for a given alert, will a screen that's no longer showing it ever get cleared? I think if we're always `union`-ing (or adding to the list in the previous list-based implementation), then things will never get removed.

Currently:
```ex
# t = 0
put_data("1", ["A"])
put_data("2", ["A"])
#=> cache holds %{"A" => MapSet["1", "2"]}

# t = 20
put_data("1", [])
put_data("2", ["A"])
#=> cache holds %{"A" => MapSet["1", "2"]}

# and every 20 seconds for the foreseeable future...
put_data("1", [])
put_data("2", ["A"])
#=> cache still holds %{"A" => MapSet["1", "2"]} because prev set is continually unioned with new set before the "A" alert key expires
```

I think this is missing the part of the implementation where each screen ID under an alert gets paired with a timestamp of when it was added. So the struct typing would now be:
```ex
%__MODULE__{
  ...
  screens_by_alert: %{
    alert_id() => %{screen_ids: list({screen_id(), timestamp :: integer()}), timer_reference: reference()}
  },
}
```

and whenever we update the list, we also clear any screen IDs that are out of date.

See "screen IDs listed under a `screens_by_alert.*` key" in the [tech spec](https://github.com/mbta/screens/blob/master/docs/tech_specs/screens_by_alert/tech_spec.md#timers-and-ttls)

Reply:
So,assuming a screen ID TTL of 30 sec for this example, it would look like:
```ex
# t = 0
put_data("1", ["A"])
put_data("2", ["A"])
#=> cache holds %{"A" => [{"1", 0}, {"2", 0}]}

# t = 20
put_data("1", [])
put_data("2", ["A"])
#=> cache holds %{"A" => [{"1", 0}, {"2", 0}, {"2", 20}]}

# t = 40
put_data("1", [])
put_data("2", ["A"])
#=> cache holds %{"A" => [{"2", 20}, {"2", 40}]}
# values from more than 30 seconds ago have been cleared
```

Reply:
Great catch. I think I have what we need there now.